### PR TITLE
[ResourceIsolation ] add driver dispatcher for workgroup

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Problem Summary：
- Implement `DriverQueueWithWorkGroup`, which dispatch drivers on multiple workgroups according to the algorithm inspired by Linux CFS.
- Add a new thread pool `wg_driver_dispatcher`, where drivers run when `enable_resource_group` is true.
- Add `DispatcherOwnerManager` to label each dispatcher threads belonging to which workgroups.
   If the dispatcher runs the driver of other workgroups, and its owner workgroup has coming driver, it will yield 20ms instead of 100ms.

### CODE REVIEW

**Review by satanson — DISMISSED** [review]

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 1]
```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
...
+    // QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher is identical to put_back.
+    put_back(driver);
+}
+
+void QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) {
+    // QuerySharedDriverQueueWithoutLock::put_back_from_dispatcher is identical to put_back.
+    put_back(drivers);
+}
+
+StatusOr<DriverRawPtr> QuerySharedDriverQueueWithoutLock::take(int dispatcher_id) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:123]: Any ut for these code?

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 2]
```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
...
+        if (_is_closed) {
+            return Status::Cancelled("Shutdown");
+        }
+    }
+
+    // Try to take driver from any owner workgroup first.
+    auto wg = _find_min_owner_wg(dispatcher_id);
+    if (wg == nullptr) {
+        // All the owner workgroups don't have ready drivers, so select the other workgroup.
+        wg = _find_min_wg();
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:215]: Could the returned wg is nullptr?

File: be/src/exec/pipeline/pipeline_driver_queue.h [hunk 3]
```diff
@@ -59,16 +73,112 @@ class QuerySharedDriverQueue : public FactoryMethod<DriverQueue, QuerySharedDriv
...
+    }
+    ~QuerySharedDriverQueueWithoutLock() override = default;
+    void close() override {}
+
+    void put_back(const DriverRawPtr driver) override;
+    void put_back(const std::vector<DriverRawPtr>& drivers) override;
+    void put_back_from_dispatcher(const DriverRawPtr driver) override;
+    void put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) override;
+
+    // return nullptr if queue is closed;
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/pipeline_driver_queue.h:117]: What's the return value if queue is closed? Why not a `Status::QueueClosed()` ?

File: be/src/exec/workgroup/work_group.cpp [hunk 4]
```diff
@@ -69,14 +69,36 @@ void WorkGroup::init() {
...
+    if (sum_cpu_runtime_ns == 0) {
+        return 0;
+    }
+    return static_cast<double>(real_runtime_ns()) / sum_cpu_runtime_ns;
+}
+
+WorkGroupManager::WorkGroupManager()
+        : _driver_dispatcher_owner_manager(std::make_unique<DispatcherOwnerManager>(
+                  config::pipeline_exec_thread_pool_thread_num > 0 ? config::pipeline_exec_thread_pool_thread_num
+                                                                   : std::thread::hardware_concurrency())),
```
Comment (mofeiatwork) [inline be/src/exec/workgroup/work_group.cpp:90]: If this interface aware of the document environment ? What if BE runs in a docker, whose CPU is limited ?

File: be/src/exec/pipeline/pipeline_driver_queue.cpp (hunk 2 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/pipeline_driver_queue.cpp:215]: @mofeiatwork  No, if the code arrives here after breaking `while (_ready_wgs.empty()) {}`, `_ready_wgs` must contain at least wg.

File: be/src/runtime/exec_env.cpp [hunk 5]
```diff
@@ -153,9 +154,19 @@ Status ExecEnv::_init(const std::vector<StorePath>& store_paths) {
                             .set_max_queue_size(1000)
                             .set_idle_timeout(MonoDelta::FromMilliseconds(2000))
                             .build(&driver_dispatcher_thread_pool));
-    _driver_dispatcher = new pipeline::GlobalDriverDispatcher(std::move(driver_dispatcher_thread_pool));
+    _driver_dispatcher = new pipeline::GlobalDriverDispatcher(std::move(driver_dispatcher_thread_pool), false);
     _driver_dispatcher->initialize(max_thread_num);
 
+    std::unique_ptr<ThreadPool> wg_driver_dispatcher_thread_pool;
+    RETURN_IF_ERROR(ThreadPoolBuilder("pip_wg_dispatcher") // pipeline dispatcher
```
Comment (mofeiatwork) [inline be/src/runtime/exec_env.cpp:161]: Is it safe to destroy the object if `_init` return error halfway?

File: be/src/exec/pipeline/pipeline_driver_queue.h (hunk 3 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/pipeline_driver_queue.h:117]: This comment already exists before this PR, and is a little misleading. Let me update this comment.
`QuerySharedDriverQueue::take()` and `DriverQueueWithWorkGroup` returns `Status::Cancelled("Shutdown")`, if queue is closed. And `QuerySharedDriverQueueWithoutLock::take()` always return non-nullable value.

File: be/src/exec/workgroup/work_group.cpp (hunk 4 above)
Comment (ZiheLiu) [reply be/src/exec/workgroup/work_group.cpp:90]: `hardware_concurrency` in Linux invokes `sysconf(_SC_NPROCESSORS_ONLN)`, which reads `/sys/devices/system/cpu/online`. If docker uses `cpuset.cpus` to limit cpu, this interface cannot be aware of this.

`hardware_concurrency` is used in lots of places in our codes. And the aim of code here is  to get the num threads of `driver_dispatcher` which is created as follows:
```C++
    auto num_threads = std::thread::hardware_concurrency();
    if (config::pipeline_exec_thread_pool_thread_num > 0) {
        num_threads = config::pipeline_exec_thread_pool_thread_num;
    }

    _driver_dispatcher = new pipeline::GlobalDriverDispatcher(num_threads, ....);
```

Therefore, maybe we could uses some plugin to modify the system file of container, like [LXCFS](https://github.com/lxc/lxcfs). Or change another way to get the number of cores in the future PR.

File: be/src/runtime/exec_env.cpp (hunk 5 above)
Comment (ZiheLiu) [reply be/src/runtime/exec_env.cpp:161]: `main()` calls `ExecEnv::init()`, but doesn't check its returned status. Let me update it.

File: be/src/exec/workgroup/work_group.cpp (hunk 4 above)
Comment (mofeiatwork) [reply be/src/exec/workgroup/work_group.cpp:90]: Got it. Thanks.

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 6]
```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
...
+    int queue_idx = -1;
+    double target_accu_time = 0;
+    DriverRawPtr driver_ptr;
+
+    for (int i = 0; i < QUEUE_SIZE; ++i) {
+        // we just search for queue has element
+        if (!_queues[i].queue.empty()) {
+            double local_target_time = _queues[i].accu_time_after_divisor();
+            // if this is first queue that has element, we select it;
+            // else we choose queue that the execution time is less sufficient,
```
Comment (liuyehcf) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:134]: Maybe this explanation can be simplify. E.g. find the queue with smallest execution time (placed before for loop)

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 7]
```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
...
+    }
+
+    // Try to take driver from any owner workgroup first.
+    auto wg = _find_min_owner_wg(dispatcher_id);
+    if (wg == nullptr) {
+        // All the owner workgroups don't have ready drivers, so select the other workgroup.
+        wg = _find_min_wg();
+    }
+
+    if (wg->driver_queue()->size() == 1) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:218]: What this condition means?

File: be/src/exec/pipeline/pipeline_driver_queue.cpp [hunk 8]
```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
...
+        if constexpr (!from_dispatcher) {
+            auto* min_wg = _find_min_wg();
+            if (min_wg != nullptr) {
+                int64_t origin_real_runtime_ns = wg->real_runtime_ns();
+                int64_t new_vruntime_ns =
+                        std::min(min_wg->vruntime_ns() - _ideal_runtime_ns(wg) / 2,
+                                 min_wg->vruntime_ns() * int64_t(min_wg->cpu_limit()) / int64_t(wg->cpu_limit()));
+                wg->set_vruntime_ns(std::max(wg->vruntime_ns(), new_vruntime_ns));
+                int64_t diff_real_runtime_ns = wg->real_runtime_ns() - origin_real_runtime_ns;
+                workgroup::WorkGroupManager::instance()->increment_cpu_runtime_ns(diff_real_runtime_ns);
```
Comment (liuyehcf) [inline be/src/exec/pipeline/pipeline_driver_queue.cpp:261]: Please add some comments, I'm kind of confused about these math expressions

File: be/src/exec/workgroup/work_group.h [hunk 9]
```diff
@@ -131,13 +145,40 @@ class WorkGroup {
...
+    DispatcherOwnerManager(const DispatcherOwnerManager&) = delete;
+    DispatcherOwnerManager& operator=(const DispatcherOwnerManager&) = delete;
+    DispatcherOwnerManager(DispatcherOwnerManager&&) = delete;
+    DispatcherOwnerManager& operator=(DispatcherOwnerManager&&) = delete;
+
+    std::shared_ptr<WorkGroupPtrSet> get_owners(int dispatcher_id) const {
+        return _dispatcher_id2owner_wgs[_index % 2][dispatcher_id];
+    }
+
+    void reassign_to_wgs(const std::unordered_map<int128_t, WorkGroupPtr>& workgroups, int sum_cpu_limit);
```
Comment (liuyehcf) [inline be/src/exec/workgroup/work_group.h:171]: add comment

File: be/src/exec/pipeline/pipeline_driver_queue.cpp (hunk 6 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/pipeline_driver_queue.cpp:134]: OK, although this comment has already existed before this PR in `QuerySharedDriverQueue`.

**Review by laotan332 — DISMISSED** [review]

**Review by satanson — APPROVED** [review]

**Review by laotan332 — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**